### DISCUSSION TITLE
What are the best options for persistent data storage with BEX?

### DISCUSSION THREAD

**Discussion by mYnDstrEAm** [quasarframework/quasar#9932] [CLI - BEX mode]
One option (that I implemented) would be storing it in `chrome.storage.local`.
However, I'm not sure if this would be a good choice in terms of privacy (vulnerabilities and other software that may access that storage), how reliable and robust that storage is (data shouldn't get lost with updates, cleaning the browser, code run by other extensions or browser crashes) and how well other apps can read/write the same data (like a Qt desktop app that uses the same data as the BEX).

Which other options are there for persistent, cross-session data-storage with BEX and which would be the best option in terms of these 3 criteria / in your view?

For example: is it possible to easily store the data in a nosql database or a sqlite database file instead? Maybe even something like [this](https://github.com/Thomaash/vuex-ltm) that can be combined with Vuex for persistent storage? Are there any example projects of such BEX data persistence?

I haven't looked into browser extensions and quasar BEX development much so far and it needs to work in Firefox, Chrome and other major browsers.

**Comment by ontwikkelfabriek** [comment]
@mYnDstrEAm 

> One option (that I implemented) would be storing it in chrome.storage.local. However, I'm not sure if this would be a good choice in terms of privacy (vulnerabilities and other software that may access that storage)

That is probably the most reliable / robust / recommended  way of persisting **non-confidential** data in BEX. For persisting confidential / privacy sensitive data I found this thread that has useful info how to store data securely in a remote database in BEX: 
https://stackoverflow.com/questions/12921562/how-do-i-securely-update-a-remote-database-from-inside-a-chrome-extension

> data shouldn't get lost with updates, cleaning the browser, code run by other extensions or browser crashes

Google will sync/persist the data when a user is logged in and has sync enabled:

_When using storage.sync, the stored data will automatically be synced to any Chrome browser that the user is logged into, provided the user has sync enabled._

_When Chrome is offline, Chrome stores the data locally. The next time the browser is online, Chrome syncs the data. Even if a user disables syncing, storage.sync will still work. In this case, it will behave identically to storage.local._

See:
https://developer.chrome.com/docs/extensions/reference/storage/#usage

> ... can be combined with Vuex for persistent storage?

I would use this Vuex plugin instead to `persist` data in Vuex:
https://github.com/championswimmer/vuex-persist




BTW I would also take a look at what Firefox doc has to say about persisting data in extension:
https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/storage

**Reply by mYnDstrEAm** [reply]
Thanks, that's very useful. I'm not interested in storing things in remote databases or syncing with Google but only storage in a **local database**.  
`vuex-persist` looks interesting but it doesn't support local databases – I already store in localStorage but would like to change it from that to a local database. I found `vuex-shared-mutations` for syncing that Vuex store across tabs (and it looks like there are also alternatives to it)
